libs/auth/src/openssl.ts

```diff
@@ -434,3 +414,109 @@ export async function createCert(input: CreateCertInput): Promise<Buffer> {
     : [scriptPath, scriptInput];
   return runCommand(command);
 }
+
+//
+// Message signing functions
+//
+
+/**
+ * The input to signMessage. File keys should be in PEM format.
+ */
+export interface SignMessageInput {
+  /**
+   * We use streaming APIs to process the message since the message could be large and infeasible
+   * to store in memory (i.e. in a Buffer), e.g. the message might include the contents of a
+   * multi-GB zip file.
+   *
+   * A seeming alternative would be to require a file path since the openssl command streams under
+   * the hood, but many of our messages will include the contents of a file preceded by a custom
+   * prefix, so there is no file path to provide (unless a temporary file is created, which too
+   * would have to be done with streaming APIs).
+   */
+  message: Stream;
+  signingPrivateKey: FileKey | TpmKey;
+}
+
+/**
+ * The input to signMessage, excluding the message
+ */
+export type SignMessageInputExcludingMessage = Omit<
+  SignMessageInput,
+  'message'
+>;
+
+const SignMessageInputExcludingMessageSchema: z.ZodSchema<SignMessageInputExcludingMessage> =
+  z.object({
+    signingPrivateKey: z.union([FileKeySchema, TpmKeySchema]),
+  });
+
+/**
+ * Parses a JSON string as a SignMessageInputExcludingMessage, throwing an error if parsing fails
+ */
+export function parseSignMessageInputExcludingMessage(
+  value: string
+): SignMessageInputExcludingMessage {
+  return unsafeParse(SignMessageInputExcludingMessageSchema, JSON.parse(value));
+}
+
+/**
+ * A helper for signMessage. Digitally signs data from the standard input using the specified
+ * private key.
+ */
+export async function signMessageHelper({
+  signingPrivateKey,
+}: SignMessageInputExcludingMessage): Promise<Buffer> {
+  const signParams: OpensslParam[] = (() => {
+    switch (signingPrivateKey.source) {
+      case 'file': {
+        return ['-sign', signingPrivateKey.path];
+      }
+      case 'tpm': {
+        return [
+          '-sign',
+          TPM_KEY_ID,
+          '-keyform',
+          'engine',
+          '-engine',
+          OPENSSL_TPM_ENGINE_NAME,
+          '-passin',
+          `pass:${TPM_KEY_PASSWORD}`,
+        ];
+      }
+      /* istanbul ignore next: Compile-time check for completeness */
+      default: {
+        throwIllegalValue(signingPrivateKey, 'source');
+      }
+    }
+  })();
+  return await openssl(['dgst', '-sha256', ...signParams], {
+    stdin: process.stdin,
+  });
+}
+
+/**
+ * Digitally signs a message using the specified private key.
+ *
+ * Uses a slightly roundabout control flow for permissions purposes:
+ *
+ * signMessage() --> ../src/intermediate-scripts/sign-message --> signMessageHelper()
+ *
+ * When using TPM keys in production, we need sudo access. Creating the intermediate sign-message
+ * script allows us to grant password-less sudo access to the relevant system users for just that
+ * operation, instead of having to grant password-less sudo access more globally.
+ */
```

Comment:
I'm using the same pattern that I landed on for `createCert`:

https://github.com/votingworks/vxsuite/blob/417f9242cfba32b266911265e0ef5faf7e6dad12/libs/auth/src/openssl.ts#L452-L464

Just as I did for `createCert`, I'll be opening a PR in vxsuite-complete-system for the permissions update:

https://github.com/votingworks/vxsuite-complete-system/blob/a6f272bc74684807c87d80be06926550fe61ab4a/config/sudoers#L33